### PR TITLE
 Fix TestEq, ConstrainEq, and CondSelect for JubjubPoint inputs

### PR DESCRIPTION
**Pull request by MicroProofs**
## Description
  - The in-circuit handlers for `TestEq`, `ConstrainEq`, and `CondSelect` force-cast operands to `AssignedNative`, crashing on `JubjubPoint` inputs
  - Extract polymorphic helpers into `ir_instructions/eq.rs` and `ir_instructions/cond_select.rs` that handle both `Native` and `JubjubPoint` via coordinate decomposition
  - Add unit tests and integration proof tests (including negative cases)

  ## Test plan
  - [x] `cargo test -p midnight-zkir-v3` — 41 tests pass (15 unit + 26 integration)
  - [x] `cargo clippy -p midnight-zkir-v3 --lib -- -D warnings` — clean
  - [x] Full workspace via Docker/nix CI shell — 559 passed, 0 failed (default features)
  - [x] Full workspace tests via Docker/nix CI shell — 615 passed, 0 failed (`--all-features`)
  - [x] Existing `test_immediate_add_and_cond_select` still passes (native CondSelect regression check)
## Sanity Checklist

This PR:
- [ ] contains changes to transaction behaviour [^1]
- [ ] contains changes to architecture [^1]
- [ ] contains breaking API changes [^1][^2]
- [ ] contains data format changes [^1]
- [x] contains circuit behaviour changes [^3]
- [ ] requires a backport to prior versions
- [x] is primarily authored by AI
- [x] I have self-reviewed the PR diff
- [x] changelog and package versions have been amended, or don't require it
- [] has ignored the checklist

### CODE REVIEW

File: zkir-v3/src/ir_instructions/eq.rs [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+
+/// Tests off-circuit whether the given inputs are equal.
+/// Equality testing is supported on:
+///   - `Native`
+///   - `JubjubPoint`
+///
+/// # Errors
+///
+/// This function results in an error if the input types are not supported.
+pub fn test_eq_offcircuit(a: &IrValue, b: &IrValue) -> Result<IrValue, anyhow::Error> {
```
Comment (JosephDenman) [inline zkir-v3/src/ir_instructions/eq.rs]: Is this function ever used? Should it be used in `preprocess`?

File: zkir-v3/src/ir_instructions/eq.rs [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+/// Constrains off-circuit the given inputs to be equal.
+/// Equality constraint is supported on:
+///   - `Native`
+///   - `JubjubPoint`
+///
+/// # Errors
+///
+/// This function results in an error if the inputs are not equal or the types
+/// are not supported.
+pub fn constrain_eq_offcircuit(a: &IrValue, b: &IrValue) -> Result<(), anyhow::Error> {
```
Comment (JosephDenman) [inline zkir-v3/src/ir_instructions/eq.rs]: Is this function ever used? Should it be used in preprocess?

File: zkir-v3/src/ir_instructions/cond_select.rs [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+/// If `bit` is true, returns `a`; otherwise returns `b`.
+///
+/// Supported on:
+///   - `Native`
+///   - `JubjubPoint`
+///
+/// # Errors
+///
+/// This function results in an error if `a` and `b` have different types.
+pub fn cond_select_offcircuit(bit: bool, a: &IrValue, b: &IrValue) -> Result<IrValue, anyhow::Error> {
```
Comment (JosephDenman) [inline zkir-v3/src/ir_instructions/cond_select.rs]: Is this function ever used? Should it be used in preprocess?

Comment (JosephDenman) [reply zkir-v3/src/ir_instructions/cond_select.rs]: Also, this function can never fail. It doesn't need to return `Result`, right?

**Review by JosephDenman — DISMISSED** [review]
Looks like a few unused functions. Also be sure to run `cargo fmt` before you merge.

File: zkir-v3/src/ir_instructions/eq.rs (hunk 2 above)
Comment (MicroProofs) [reply zkir-v3/src/ir_instructions/eq.rs]: my mistake it's used in preprocess now

**Comment by MicroProofs** [comment]
> Looks like a few unused functions. Also be sure to run `cargo fmt` before you merge.

Ran clippy and set up a hook on my end so it autoruns

**Comment by kmillikin** [comment]
@miguel-ambrona, you should take a look at this.

File: zkir-v3/src/ir_instructions/cond_select.rs [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+/// Conditionally selects off-circuit between two values.
+/// If `bit` is true, returns `a`; otherwise returns `b`.
+///
+/// Supported on:
+///   - `Native`
+///   - `JubjubPoint`
+///
+/// # Errors
+///
+/// This function results in an error if `a` and `b` have different types.
```
Comment (miguel-ambrona) [inline zkir-v3/src/ir_instructions/cond_select.rs]: This doc comment is not aligned with the implementation.

File: zkir-v3/src/ir_instructions/select.rs [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+///   - `Native`
+///   - `JubjubPoint`
+///
+/// # Errors
+///
+/// This function results in an error if the input types are not supported.
+pub fn cond_select_incircuit(
+    std_lib: &ZkStdLib,
+    layouter: &mut impl Layouter<F>,
+    bit: &AssignedBit<F>,
```
Comment (miguel-ambrona) [inline zkir-v3/src/ir_instructions/select.rs:58]: I would rename this argument: `bit` -> `cond`.

File: zkir-v3/src/ir_instructions/cond_select.rs [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    b: &CircuitValue,
+) -> Result<CircuitValue, plonk::Error> {
+    use CircuitValue::*;
+    match (a, b) {
+        (Native(x), Native(y)) => Ok(Native(std_lib.select(layouter, bit, x, y)?)),
+        (JubjubPoint(p), JubjubPoint(q)) => {
+            let jub = std_lib.jubjub();
+            let rx = std_lib.select(layouter, bit, &jub.x_coordinate(p), &jub.x_coordinate(q))?;
+            let ry = std_lib.select(layouter, bit, &jub.y_coordinate(p), &jub.y_coordinate(q))?;
+            Ok(JubjubPoint(jub.point_from_coordinates(layouter, &rx, &ry)?))
```
Comment (miguel-ambrona) [inline zkir-v3/src/ir_instructions/cond_select.rs]:
```suggestion
            Ok(JubjubPoint(std_lib.jubjub().select(layouter, bit, p, q)?))
```

File: zkir-v3/src/ir_instructions/eq.rs [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
```
Comment (miguel-ambrona) [inline zkir-v3/src/ir_instructions/eq.rs]: I would have 2 different files:
 - test_eq (equality checks)
 - constrain_eq (equality assertions)

File: zkir-v3/src/ir_instructions/eq.rs [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+            Ok(Native(std_lib.convert(layouter, &bit)?))
+        }
+        (JubjubPoint(p), JubjubPoint(q)) => {
+            let jub = std_lib.jubjub();
+            let bit_x = std_lib.is_equal(layouter, &jub.x_coordinate(p), &jub.x_coordinate(q))?;
+            let bit_y = std_lib.is_equal(layouter, &jub.y_coordinate(p), &jub.y_coordinate(q))?;
+            let nx = std_lib.convert(layouter, &bit_x)?;
+            let ny = std_lib.convert(layouter, &bit_y)?;
+            Ok(Native(std_lib.mul(layouter, &nx, &ny, None)?))
+        }
```
Comment (miguel-ambrona) [inline zkir-v3/src/ir_instructions/eq.rs]:
```suggestion
        (Native(x), Native(y)) => std_lib.is_equal(layouter, x, y),
        (JubjubPoint(p), JubjubPoint(q)) => std_lib.jubjub().is_equal(layouter, p, q),
```

File: zkir-v3/src/ir_instructions/eq.rs [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+///
+/// # Errors
+///
+/// This function results in an error if the input types are not supported.
+pub fn test_eq_incircuit(
+    std_lib: &ZkStdLib,
+    layouter: &mut impl Layouter<F>,
+    a: &CircuitValue,
+    b: &CircuitValue,
+) -> Result<CircuitValue, plonk::Error> {
```
Comment (miguel-ambrona) [inline zkir-v3/src/ir_instructions/eq.rs]: It would be good to introduce case `Bit` or `Bool` in `CircuitValue`. Until then:

```suggestion
) -> Result<AssignedBit<F>, plonk::Error> {
```

File: zkir-v3/src/ir_instructions/eq.rs [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+    b: &CircuitValue,
+) -> Result<(), plonk::Error> {
+    use CircuitValue::*;
+    match (a, b) {
+        (Native(x), Native(y)) => std_lib.assert_equal(layouter, x, y),
+        (JubjubPoint(p), JubjubPoint(q)) => {
+            let jub = std_lib.jubjub();
+            std_lib.assert_equal(layouter, &jub.x_coordinate(p), &jub.x_coordinate(q))?;
+            std_lib.assert_equal(layouter, &jub.y_coordinate(p), &jub.y_coordinate(q))
+        }
```
Comment (miguel-ambrona) [inline zkir-v3/src/ir_instructions/eq.rs]:
```suggestion
        (JubjubPoint(p), JubjubPoint(q)) => std_lib.jubjub().assert_equal(layouter, p, q),
```

File: CHANGELOG_zkir.md [hunk 11]
```diff
@@ -1,5 +1,7 @@
 # `zkir` Changelog
 
+- bugfix: make `TestEq`, `ConstrainEq`, and `CondSelect` polymorphic over `Native` and `JubjubPoint` types (previously crashed on `JubjubPoint` inputs)
```
Comment (miguel-ambrona) [inline CHANGELOG_zkir.md]: Why "bugfix". This was not a bug.

File: zkir-v3/src/ir_instructions/cond_select.rs [hunk 12]
```diff
@@ -28,11 +28,8 @@ use crate::{
 ///   - `Native`
 ///   - `JubjubPoint`
 ///
-/// # Errors
-///
-/// This function results in an error if `a` and `b` have different types.
-pub fn cond_select_offcircuit(bit: bool, a: &IrValue, b: &IrValue) -> Result<IrValue, anyhow::Error> {
-    Ok(if bit { a.clone() } else { b.clone() })
+pub fn cond_select_offcircuit(bit: bool, a: &IrValue, b: &IrValue) -> IrValue {
```
Comment (miguel-ambrona) [inline zkir-v3/src/ir_instructions/cond_select.rs]: This function should fail if given inputs of different types.

**Review by miguel-ambrona — CHANGES_REQUESTED** [review]
Why are we adding these functions? I was intentionally postponing this as we still do not have `Bool` in `CircuitValue` and we still did not confirm that these operations would be polymorphic, are they in Compact yet?

I would not have the "offcircuit" version of trivial functions like equality assertions, equality checks.
Maybe for `select` it makes sense to have the offcircuit version, but only in order to fail if the input types are not equal.
Btw, why not `cond_select` -> `select`? I'm not sure what the prefix "cond" is adding here.

About the tests added on tests/proofs. Can they be aggregated in the same test function? At least the positive tests, they are really verbose.

File: zkir-v3/src/ir_instructions/add.rs [hunk 13]
```diff
@@ -35,7 +35,7 @@ pub fn add_offcircuit(x: &IrValue, y: &IrValue) -> Result<IrValue, anyhow::Error
     match (x, y) {
         (Native(a), Native(b)) => Ok(Native(*a + *b)),
         (JubjubPoint(p), JubjubPoint(q)) => Ok(JubjubPoint(p + q)),
-        _ => Err(anyhow::anyhow!(
+        (Native(_), JubjubPoint(_)) | (JubjubPoint(_), Native(_)) => Err(anyhow::anyhow!(
```
Comment (miguel-ambrona) [inline zkir-v3/src/ir_instructions/add.rs]: It is good to avoid underscores, but in this case it is justified.
If the number of types grows, this will become a nightmare to maintain.

File: zkir-v3/src/ir_instructions/add.rs [hunk 14]
```diff
@@ -67,11 +67,13 @@ pub fn add_incircuit(
             let r = std_lib.jubjub().add(layouter, p, q)?;
             Ok(JubjubPoint(r))
         }
-        _ => Err(plonk::Error::Synthesis(format!(
-            "Unsupported addition: {:?} + {:?}",
-            x.get_type(),
-            y.get_type()
-        ))),
+        (Native(_), JubjubPoint(_)) | (JubjubPoint(_), Native(_)) => {
```
Comment (miguel-ambrona) [inline zkir-v3/src/ir_instructions/add.rs]: Again, an `_` is preferable here.

File: zkir-v3/src/ir_instructions/constrain_eq.rs [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+        (Native(_), Native(_)) | (JubjubPoint(_), JubjubPoint(_)) => Err(anyhow::anyhow!(
+            "Equality constraint failed: {:?} != {:?}",
+            a,
+            b
+        )),
+        (Native(_), JubjubPoint(_)) | (JubjubPoint(_), Native(_)) => Err(anyhow::anyhow!(
+            "Unsupported constrain_eq: {:?} == {:?}",
+            a.get_type(),
+            b.get_type()
+        )),
```
Comment (miguel-ambrona) [inline zkir-v3/src/ir_instructions/constrain_eq.rs]: Use underscores.

File: zkir-v3/src/ir_instructions/constrain_eq.rs [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+    std_lib: &ZkStdLib,
+    layouter: &mut impl Layouter<F>,
+    a: &CircuitValue,
+    b: &CircuitValue,
+) -> Result<(), plonk::Error> {
+    use CircuitValue::*;
+    match (a, b) {
+        (Native(x), Native(y)) => std_lib.assert_equal(layouter, x, y),
+        (JubjubPoint(p), JubjubPoint(q)) => std_lib.jubjub().assert_equal(layouter, p, q),
+        (Native(_), JubjubPoint(_)) | (JubjubPoint(_), Native(_)) => {
```
Comment (miguel-ambrona) [inline zkir-v3/src/ir_instructions/constrain_eq.rs]: Underscore.

File: zkir-v3/src/ir_instructions/eq.rs [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+///
+/// # Errors
+///
+/// This function results in an error if the input types are not supported.
+pub fn test_eq_offcircuit(a: &IrValue, b: &IrValue) -> Result<IrValue, anyhow::Error> {
+    use IrValue::*;
+    match (a, b) {
+        (Native(x), Native(y)) => Ok(IrValue::Native((x == y).into())),
+        (JubjubPoint(p), JubjubPoint(q)) => Ok(IrValue::Native((p == q).into())),
+        (Native(_), JubjubPoint(_)) | (JubjubPoint(_), Native(_)) => Err(anyhow::anyhow!(
```
Comment (miguel-ambrona) [inline zkir-v3/src/ir_instructions/eq.rs]: Underscore.

File: zkir-v3/src/ir_instructions/eq.rs [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+    std_lib: &ZkStdLib,
+    layouter: &mut impl Layouter<F>,
+    a: &CircuitValue,
+    b: &CircuitValue,
+) -> Result<AssignedBit<F>, plonk::Error> {
+    use CircuitValue::*;
+    match (a, b) {
+        (Native(x), Native(y)) => std_lib.is_equal(layouter, x, y),
+        (JubjubPoint(p), JubjubPoint(q)) => std_lib.jubjub().is_equal(layouter, p, q),
+        (Native(_), JubjubPoint(_)) | (JubjubPoint(_), Native(_)) => {
```
Comment (miguel-ambrona) [inline zkir-v3/src/ir_instructions/eq.rs]: Underscore.

File: zkir-v3/src/ir_instructions/select.rs [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+    a: &CircuitValue,
+    b: &CircuitValue,
+) -> Result<CircuitValue, plonk::Error> {
+    use CircuitValue::*;
+    match (a, b) {
+        (Native(x), Native(y)) => Ok(Native(std_lib.select(layouter, bit, x, y)?)),
+        (JubjubPoint(p), JubjubPoint(q)) => {
+            Ok(JubjubPoint(std_lib.jubjub().select(layouter, bit, p, q)?))
+        }
+        (Native(_), JubjubPoint(_)) | (JubjubPoint(_), Native(_)) => {
```
Comment (miguel-ambrona) [inline zkir-v3/src/ir_instructions/select.rs]: Underscore.

File: zkir-v3/src/ir_instructions/constrain_eq.rs [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+                    "Equality constraint failed: {a:?} != {b:?}"
+                ))
+            }
+        }
+        _ => Err(anyhow::anyhow!(
+            "Unsupported constrain_eq: {:?} == {:?}",
+            a.get_type(),
+            b.get_type()
+        )),
+    }
```
Comment (miguel-ambrona) [inline zkir-v3/src/ir_instructions/constrain_eq.rs]: If we do it this way, no more edits will be necessary when new types come:


```suggestion
    if a.get_type() != b.get_type() {
        return Err(anyhow::anyhow!(
            "Unsupported constrain_eq: {:?} == {:?}",
            a.get_type(),
            b.get_type()
        ));
    }

    if a != b {
        return Err(anyhow::anyhow!(
            "Equality constraint failed: {a:?} != {b:?}"
        ));
    }

    Ok(())
```

File: zkir-v3/src/ir_instructions/eq.rs [hunk 21]
```diff
@@ -0,0 +1,96 @@
...
+
+/// Tests off-circuit whether the given inputs are equal.
+/// Equality testing is supported on:
+///   - `Native`
+///   - `JubjubPoint`
+///
+/// # Errors
+///
+/// This function results in an error if the input types are not supported.
+pub fn test_eq_offcircuit(a: &IrValue, b: &IrValue) -> Result<IrValue, anyhow::Error> {
```
Comment (miguel-ambrona) [inline zkir-v3/src/ir_instructions/eq.rs]: I would return `Result<bool, anyhow::Error>`.

**Review by miguel-ambrona — DISMISSED** [review]

**Review by miguel-ambrona — DISMISSED** [review]

**Comment by MicroProofs** [comment]
Depends on #278

**Review by tkerber — DISMISSED** [review]

**Review by tkerber — DISMISSED** [review]

**Review by tkerber — APPROVED** [review]
